bigml/externalconnectorhandler.py

```diff
@@ -0,0 +1,141 @@
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+"""Base class for whizzml script external connectors' REST calls
+
+   https://bigml.com/api/externalconnectors
+
+"""
+import os
+try:
+    import simplejson as json
+except ImportError:
+    import json
+
+
+from bigml.resourcehandler import ResourceHandler
+from bigml.resourcehandler import check_resource_type, \
+    get_external_connector_id, get_resource_type, check_resource
+from bigml.constants import EXTERNAL_CONNECTOR_PATH, TINY_RESOURCE, \
+    EXTERNAL_CONNECTION_ATTRS
+
+
+def get_env_connection_info():
+    """Retrieves the information to use in the external connection from
+    environment variables.
+
+    """
+    # try to use environment variables values
+    connection_info = {}
+    for external_key in EXTERNAL_CONNECTION_ATTRS.keys():
+        if os.environ.get(external_key):
+            connection_info.update( \
+                {EXTERNAL_CONNECTION_ATTRS[external_key]:
+                 os.environ.get(external_key)})
+    return connection_info
+
+
+class ExternalConnectorHandler(ResourceHandler):
+    """This class is used by the BigML class as
+       a mixin that provides the external connectors' REST calls. It should not
+       be instantiated independently.
+
+    """
+    def __init__(self):
+        """Initializes the ExternalConnectorHandler. This class is intended to
+           be used as a mixin on ResourceHandler, that inherits its
+           attributes and basic method from BigMLConnection, and must not be
+           instantiated independently.
+
+        """
+        self.external_connector_url = self.url + EXTERNAL_CONNECTOR_PATH
+
+    def create_external_connector(self, connection_info, args=None,
+                                  wait_time=3, retries=10):
+        """Creates an external connections from a dictionary containing the
+        connection information.
+
+        """
+
+        create_args = {}
+        if args is not None:
+            create_args.update(args)
+
+        if connection_info is None:
+            connection_info = get_env_connection_info()
+
+        if not isinstance(connection_info, dict):
+            raise Exception("To create an external connector you need to"
+                            " provide a dictionary with the connection"
+                            " information. Please refer to the API external"
+                            " connector docs for details.")
+
+        source = connection_info.get("source", "postgresql")
```

Comment:
If the user adds connections details for a different kind of database and forgets to add the "source" field, using postgresql as a default value can be confusing, can't be?

Reply:
could be, but the thing is that the "source" attribute is compulsory (it wasn't validated at the API level, and that's why I added a default value, but now it is). We could just not use a default and let the create call fail. I rather prefer setting a default to the thing that we assume will be more commonly used. That's a guess of course, but seeing that we are using it in man y projects, I think that is sensible enough. That said, I should add a comment in the docs saying so. Would that be ok?

Reply:
Ok, great. After all, I guess that in the case we were wrong with our guess, the user will receive an error from the API

Reply:
Yes, that's exactly what will happen.